src/Transports.AspNetCore.SystemTextJson/GraphQLRequestDeserializer.cs

```diff
@@ -0,0 +1,134 @@
+    /// </remarks>
+    public class GraphQLRequestDeserializer : IGraphQLRequestDeserializer
+    {
+        private readonly JsonSerializerOptions _serializerOptions = new JsonSerializerOptions();
+
+        public GraphQLRequestDeserializer(Action<JsonSerializerOptions> configure)
+        {
+            // Add converter that deserializes Variables property
+            _serializerOptions.Converters.Add(new ObjectDictionaryConverter());
+
+            configure?.Invoke(_serializerOptions);
+        }
+
+        public async Task<GraphQLRequestDeserializationResult> DeserializeFromJsonBodyAsync(HttpRequest httpRequest, CancellationToken cancellationToken = default)
+        {
+            var bodyReader = httpRequest.BodyReader;
+
+            JsonTokenType jsonTokenType;
+            try
+            {
+                jsonTokenType = await PeekJsonTokenTypeAsync(bodyReader, cancellationToken);
+            }
+            catch (JsonException)
+            {
+                // Invalid request content, assign None so it falls through to WasSuccessful = false
+                jsonTokenType = JsonTokenType.None;
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var result = new GraphQLRequestDeserializationResult() { IsSuccessful = true };
+            switch (jsonTokenType)
+            {
+                case JsonTokenType.StartObject:
+                    result.Single = ToGraphQLRequest(
+                        await JsonSerializer.DeserializeAsync<InternalGraphQLRequest>(bodyReader.AsStream(), _serializerOptions, cancellationToken));
+                    return result;
+                case JsonTokenType.StartArray:
+                    result.Batch = (await JsonSerializer.DeserializeAsync<InternalGraphQLRequest[]>(bodyReader.AsStream(), _serializerOptions, cancellationToken))
+                        .Select(ToGraphQLRequest)
+                        .ToArray();
+                    return result;
+                default:
+                    result.IsSuccessful = false;
+                    return result;
+            }
+        }
+
+        private static async ValueTask<JsonTokenType> PeekJsonTokenTypeAsync(PipeReader reader, CancellationToken cancellationToken = default)
+        {
+            // Separate method so that we can use the ref struct
+            static bool DetermineTokenType(in ReadOnlySequence<byte> buffer, out JsonTokenType jsonToken)
+            {
+                var jsonReader = new Utf8JsonReader(buffer);
+                if (jsonReader.Read())
+                {
+                    jsonToken = jsonReader.TokenType;
+                    return true;
+                }
+                jsonToken = JsonTokenType.None;
+                return false;
+            }
+
+            while (true)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var result = await reader.ReadAsync(cancellationToken);
```

Comment:
I think that it is worth to check `result.IsCanceled` after this line.

Reply:
Done.